documentation/ArrayExpressionLoweringToFIR.md

```diff
@@ -0,0 +1,315 @@
+
+
+## Examples
+Below are illustrations of Fortran to FIR lowering based on the above design (with one node, one loop). 
+
+### Simple examples
+Fortran:
+```
+  real :: x(4), y(4)
+  y = x
+```
+Here, `x` expression should not be split into a loop.
+
+FIR:
+```
+ %1 = fir.load %x // that is the whole expression evaluation
+ // This is actually the assignment lowering
+ %fir.store %1 to %y
+```
+Question: This is more related to assignment lowering, but what will decide whether `%1`
+is has to be copied inside a temporary or not (e.g. in case of pointers that may overlap or not).
+Is it implicit in here because `%1` requires a storage later and
+FIR will determine whether `%y` can directly be used ? 
+### Outline scalar sub-expression outside of the loop
+Fortran:
+```
+  real :: x(4), y(4)
+  y = x + cos(1.)
+```
+FIR:
+```
+  %tmp = fir.alloca !fir.array<4:f32>
+  %cst = constant 1.000000e+00 : f32
+  %1 = fir.call "cos" %cst
+  fir.loop %i = 1 to 4 {
+    %2 = fir.extract_value %x, %i
+    %3 = fir.addf %2, %1
+    %4 = fir.coordinate_of %tmp, %i
+    fir.store %3 to %4 
+  }
+  %5 = fir.load %tmp // This is the expression result
+  // The assignment part is not part of the expression lowering
+  fir.store %5 to %y
+```
+### Dynamic shape example
+A more complex case where the shape is not a compile time constant.
+Fortran:
+```
+   real, allocatable :: x(:), y(:)
+   y = x + cos(1.)
+```
+FIR:
+```
+  %_,_, %1 = fir.box_dims %x
+  %cst_0 = constant 0 : i64
+  %3 = cmpi eq %1 %cst_0
+  cond_br %3 ^bnullshape ^bb1
+  ^bb1:
+    %tmp = fir.allocmem !fir.array<?:f32> %1
+    %cst = constant 1.000000e+00 : f32
+    %cosres = fir.call "cos" %cst
+    %lbx1, _, _ = fir.box_dims %x
+    %cst_1 = constant 1 : i64
+    %lbx0 = subi lbx1 %cst_1
+    fir.loop %i = 1 to %1 {
+
+      %ix = subi %i %lbx0
+      %2 = fir.coordinate_of %x, %i
+      %3 = fir.load %2
+      %4 = fir.addf %3, %cosres
+      %5 = fir.coordinate_of %tmp, %i
+      fir.store %4 to %5
+    }
+    %6 = fir.load %tmp
+    fir.freemem %tmp // are we allowed to do this so soon?
+    jmp bresult(%6)
+  ^bnullshape:
+    // Create a null shaped result.
+    %nullres = undef !fir.array<0:f32>
+    %r = fir.convert %nullres !fir.array<?:f32> // not sure we need this
+    jmp bresult(%r)
+  ^bresult(%result : !fir.array<?:f32>)
+    //%result is the expression result
+
+  // The assignment is not part of the expression lowering
+  // It is also more complex because y is an allocatable so not shown
+  // here
+```
+
+Observation: The loop is on the shape, not the actual bounds of `x`. So we need to compute the related coordinate for `x` inside the loop.
+Here, it is a bit stupid, but in `x + y`, it does not make sense to loop on `x` bounds because the bounds of y might be different, so we loop on
+the extent starting from 1.
+
+Questions: Does it matter? Could we hide this inside `fir.coordinate_of` ?
```

Comment:
Yes. The idea of `coordinate_of` on a box value is the same as an access map being applied.
```
   %0 = fir.gendims 5, 30, 5
   %1 = fir.embox %x, %0
   call @foo(%1, ...)

   func @foo(%box...) {
      %1 = fir.coordinate_of %box, 2   // return &x[10]
```

Reply:
OK, that make sens, so if we have:
```
function foo(x, y, z)
  real :: z(1:5), x(5:10), y(10:15)
  z  = x + y
end function
```
We would create `fir.box` with the local mapping with the `fir.ref<fir.array<5:f32>>` that we got and then use the boxes when accessing `x` and `y` inside the function instead of the function reference ?
This means we can create `fir.box` locally even if we do not need a full Fortran descriptor ?

Reply:
Yes. A box is a fancy reference.

Yes. Nothing prohibits using boxes whenever it makes sense to do so. :)